test/dynamo/test_misc.py

```diff
@@ -3069,6 +3070,87 @@ def guard_failures(failure):
                 "tensor 'x' size mismatch at index 0. expected 2, actual 3",
             )
 
+    def test_restore_graphstate(self):
+        # This function does some guard accumulation,
+        # and then rolls back due to control flow.
+        # The idea is that if one were printing guards as they appear,
+        # they would see this insert a guard that does not show up in the final set of
+        # guards as we rolled back from it.
+        def nested_fn(s):
+            if x[0] < 10:
+                return s * s
+            return s
+
+        def fn(x, y):
+            x = x + 1
+            y = nested_fn(y)
+            y = y + 10
+            return x * y
+
+        all_guards = []
+
+        def guard_export_print(guards):
+            nonlocal all_guards
+            all_guards.extend(guards)
+
+        opt_fn = torch._dynamo.optimize("eager", guard_export_fn=guard_export_print)(fn)
+
+        x = torch.tensor([0.5, 0.5])
+        y = torch.tensor([1.0, 1.0])
+        opt_fn(x, y)
+
+        if torch._dynamo.config.dynamic_shapes:
+            self.assertEqual(len(all_guards), 13)
+        else:
+            self.assertEqual(len(all_guards), 9)
+        for guard in all_guards:
+            # This guard was created
+            self.assertTrue(guard.name != "nested_fn.__closure__[0].cell_contents")
```

Comment:
This seems fairly sensitive to the details of how exactly we name closure cell sources. Also, prefer using assertNotEqual here.

Reply:
Sure, this is what I meant by weak test - I dont really have a good way of capturing this...

Reply:
OK, so you don't have to do it in this PR, but it would be nice if we had some synthetic guards for testing purposes. You'd have a testing only function that installs a special guard that you can then easily identify in tests, and then you could just match that directly here

Reply:
Hmm, I really like that idea.